### ISSUE TITLE
[C#] Implement new Concept Exercise: lists

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#958]
This issue describes how to implement the `lists` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Lists in [C#][lists-docs]. We'll use the list type to teach the student about generics too. We should be using the list with more than one data type to show that is possible. 

The `list` collection type was chosen as the first generic type for the following reasons:

- It is the most commonly used generic type.
- They are a common data type in many language.

## Learning objectives

- Know what generic types are.
- Know of the existence of the `List<T>` type.
- Know how a list is different from an array.
- Know how to define a list.
- Know how to add and remove elements from a list.
- Know how to access elements in a list by index.
- Know how to iterate over elements in a list.
- Know some basic list functions (like finding the index of an element in a list or sorting a list).

## Out of scope

- Generic functions.
- Generic constraints.
- Memory and performance characteristics.
- LINQ.
- Concurrency issues.
- Co-/contravariance.
- Equality.
- List resizing due to it using an array as its data type.

## Concepts

This Concepts Exercise's Concepts are:

- `lists`: know of the existence of the `List<T>` type; know how a list is different from an array; know how to define a list; know how to add and remove elements from a list; know how to access elements in a list by index; know how to iterate over elements in a list; know some basic list functions (like finding the index of an element in a list).
- `generic-types`: know what generic types are.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `foreach-loops`: know how to use a `foreach-loop` to iterate over a collection.
- `arrays`: know of the `array` collection type and that it has a fixed length.

Any data types used in this exercise (e.g. `strings`) should also be added as prerequisites.

### Hints

- [List documentation][lists-docs]: reference documentation for `List<T>`.
- [Lists tutorial][lists-tutorial]: basic tutorial on how to work with lists.

### After

- [List documentation][lists-docs]: reference documentation for `List<T>`.
- [Lists tutorial][lists-tutorial]: basic tutorial on how to work with lists.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-list]: https://github.com/exercism/v3/blob/master/reference/types/list.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[lists-docs]: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=netcore-3.1
[lists-tutorial]: https://csharp.net-tutorials.com/collections/lists/

**Comment by neiesc** [comment]
Hi, I want to implement this issue

**Comment by mikedamay** [comment]
Hi @neiesc.  I’m not sure if you’ve kept up with the v3 progress on the C# track but we’re in the process of finishing off a bunch of introductory exercises and it would great if we could include the `lists` exercise to which you are currently assigned.

If you think you will be in a position to work on this in the next couple of weeks then please let us know and we will set our expectations according.  If not, could you let us know and we’ll be happy to reassign the exercise.  We appreciate both good intentions and how life gets in the way so no worries.

If you have any uncommitted work that you think we could use or cannibalize then please let us know.

**Comment by neiesc** [comment]
Hi @mikedamay , I even started to make a structure, but I was unable to progress, in these next weeks I will be busy I am also sorry I should have signaled before.

**Comment by mikedamay** [comment]
No problem, @neiesc.  We will take it over.

**Comment by mikedamay** [comment]
closed with PR  #2000
